lib/awestruct/rack/generate.rb

```diff
@@ -0,0 +1,46 @@
+
+module Awestruct
+  module Rack
+
+    class GenerateOnAccess
+      def initialize(app)
+        @app = app
+      end
+
+      def call(env)
+        engine = ::Awestruct::Engine.instance
+
+        generate = false
+
+        req_path = env['REQUEST_PATH']
+        path = req_path
+        path = req_path + "index.html" if req_path.end_with? '/'
+
+        page = engine.site.pages_by_output_path[path]
+        if page.nil? and !req_path.end_with? '/'
+          path = req_path + "/index.html"
+          page = engine.site.pages_by_output_path[path]
+        end
+
+        generate_path = nil
+
+        if !page.nil?
+          generate_path = File.join( engine.site.config.output_dir, page.output_path )
+
+          generate = true if page.stale?
+          generate = true if path.end_with? '.html'
+          generate = true if !File.exists? generate_path
```

Comment:
Questionable.. stale? should probably be able to return true if the expected output path does not exist.